src/components/Auth/Auth.tsx

```diff
@@ -1,44 +1,74 @@
-import React from 'react';
+import React, { useState, useCallback } from 'react';
 import { Navigate, Outlet, useLocation } from 'react-router-dom';
-import api from '../../api/api';
-
-const LOCALSTORAGE_USER_NAME = 'zaaktypenbeheer_user';
+import { UserDataT } from '../../types/types';
+import { useAsync } from 'react-use';
+import { PermissionDenied } from '../../errors/errors';
+import { get, post } from '../../api/api';
 
 export interface AuthContextType {
   user: any;
-  signIn: (formData: FormData, callback: VoidFunction) => void;
-  signOut: (callback: VoidFunction) => void;
+  onSignIn: (formData: FormData, callback: VoidFunction) => void;
+  onSignOut: (callback: VoidFunction) => void;
 }
 
-let AuthContext = React.createContext<AuthContextType>(null!);
+const AuthContext = React.createContext<AuthContextType>(null!);
 
 export function AuthProvider({ children }: { children: React.ReactNode }) {
-  let [user, setUser] = React.useState<any>(localStorage.getItem(LOCALSTORAGE_USER_NAME));
+  const [user, setUser] = useState<UserDataT>(null!);
+  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
 
-  let signIn = async (formData: FormData, callback: VoidFunction) => {
-    await api.signIn(formData);
-    localStorage.setItem(LOCALSTORAGE_USER_NAME, formData.get('username') as string);
-    setUser(formData.get('username'));
-    callback();
-  };
+  const { loading } = useAsync(async () => {
+    if (isLoggedIn) return;
+    let userData;
+
+    try {
+      userData = await get(`users/me/`);
+    } catch (e) {
+      // Catch only the expected error (403), throw any other unexpected case.
+      if (!(e instanceof PermissionDenied)) throw e;
+      return;
+    }
+    setUser(userData);
+    setIsLoggedIn(true);
+  }, []);
+
+  const onSignIn = useCallback(async (formData: FormData, callback: VoidFunction) => {
+    const data = {
+      username: formData.get('username'),
+      password: formData.get('password'),
+    };
 
-  let signOut = (callback: VoidFunction) => {
-    setUser(null);
+    await get(''); // get CSRF token
+    await post('auth/login/', data);
+    const userData = await get(`users/me/`);
+
+    setUser(userData);
+    setIsLoggedIn(true);
     callback();
-  };
+  }, []);
+
+  const onSignOut = useCallback(async (callback: VoidFunction) => {
+    await post('auth/logout/', {});
 
-  let value = { user, signIn, signOut };
+    setUser(null!);
```

Comment:
What does this `!` do here (and line 14)?
On the typescript docs I found this: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator

but it doesn't make sense since here the operand is `null` :thinking:  So I guess I'm missing something :sweat_smile: